src/API.jl

```diff
@@ -742,16 +749,33 @@ function instantiate(ctx::Context; manifest::Union{Bool, Nothing}=nothing,
     pkgs = Operations.load_all_deps(ctx)
     Operations.check_registered(ctx, pkgs)
     new_git = UUID[]
+    # Handling packages tracking repos
     for pkg in pkgs
-        pkg.repo.url !== nothing || continue
+        pkg.repo.source !== nothing || continue
         sourcepath = Operations.source_path(pkg)
         isdir(sourcepath) && continue
-        # download repo at tree hash
-        push!(new_git, pkg.uuid)
-        clonepath = Types.clone_path!(ctx, pkg.repo.url)
-        tmp_source = Types.repo_checkout(ctx, clonepath, string(pkg.tree_hash))
+        ## Download repo at tree hash
+        # determine canonical form of repo source
+        if isurl(pkg.repo.source)
+            repo_source = pkg.repo.source
+        else
+            repo_source = normpath(joinpath(dirname(ctx.env.project_file), pkg.repo.source))
+        end
+        clonepath = Types.clone_path(repo_source)
+        # We try to avoid updating the clone if the target already exists
+        if tree_hash_exists(clonepath, string(pkg.tree_hash))
+            tmp_source = Types.repo_checkout(ctx, clonepath, string(pkg.tree_hash))
```

Comment:
This line and the one in the `else` below are identical so couldn't this whole thing be changed to

```jl
if !tree_hash_exists(clonepath, string(pkg.tree_hash))
    if !isurl(repo_source) && !isdir(repo_source)
        pkgerror("Expected $(err_rep(pkg)) to exist at `$(repo_source)`, but no such directory.")
    end
    Types.clone_path!(ctx, repo_source) # TODO a more direct way of updating a clone
end
tmp_source = Types.repo_checkout(ctx, clonepath, string(pkg.tree_hash))
```